From c9dc10142c0fdb06aaaa75de7200b781ddb28531 Mon Sep 17 00:00:00 2001
From: Selvavignesh Shanmugam <sshanmugam@maxlinear.com>
Date: Thu, 26 Sep 2024 20:08:40 +0800
Subject: [PATCH] WLANRTSYS-85083 Enable hostapd_cli STA disassociate cmd for
 MLO

Issue:
When the disassociate STA CLI cmd is given, it will not be handled
if it is a MLO VAP. It was done to avoid STA disconnect issue in PF
caused by PrplMesh.

Fix:
Since the PrplMesh is disabled in PF Testcases, removing this
commit: 2de0dc1ad4d
Handled the STA disassociate cmd for ML STA using its STA MLD MAC
addr or the link addr.

FIXUP: WLANRTSYS-62845 AP MLD Auth Assoc and eapol changes
---
 hostapd/ctrl_iface.c     |  2 --
 hostapd/mxl_ctrl_iface.c | 12 ------------
 src/ap/ctrl_iface_ap.c   | 21 +++++++++++++++++++++
 3 files changed, 21 insertions(+), 14 deletions(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 35de9f3aa..cbd2a0688 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -6300,11 +6300,9 @@ static int hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 	} else if (os_strncmp(buf, "DEAUTHENTICATE ", 15) == 0) {
 		if (hostapd_ctrl_iface_deauthenticate(hapd, buf + 15))
 			reply_len = -1;
-#ifndef CONFIG_IEEE80211BE_MXL_MLO
 	} else if (os_strncmp(buf, "DISASSOCIATE ", 13) == 0) {
 		if (hostapd_ctrl_iface_disassociate(hapd, buf + 13))
 			reply_len = -1;
-#endif /* CONFIG_IEEE80211BE_MXL_MLO */
 #ifdef CONFIG_TAXONOMY
 	} else if (os_strncmp(buf, "SIGNATURE ", 10) == 0) {
 		reply_len = hostapd_ctrl_iface_signature(hapd, buf + 10,
diff --git a/hostapd/mxl_ctrl_iface.c b/hostapd/mxl_ctrl_iface.c
index 64e53d565..c0c0b1fd7 100644
--- a/hostapd/mxl_ctrl_iface.c
+++ b/hostapd/mxl_ctrl_iface.c
@@ -2903,16 +2903,6 @@ static int mxl_hostapd_ctrl_iface_send_t2lm_teardown_frame(struct hostapd_data *
 	return ml_sta_send_t2lm_teardown_frame(hapd, addr);
 }
 
-static void mxl_mlo_hostapd_ctrl_disassociate(struct hostapd_data *hapd, char *buf, int *reply_len)
-{
-	if (mxl_is_mlo_enabled(hapd->conf)) {
-		*reply_len = -1;
-	} else {
-		if (hostapd_ctrl_iface_disassociate(hapd, buf))
-			*reply_len = -1;
-	}
-}
-
 #endif /* CONFIG_IEEE80211BE_MXL_MLO */
 
 static int mxl_hostapd_ctrl_iface_set_neighbor_psd(struct hostapd_data *hapd, char *buf)
@@ -3127,8 +3117,6 @@ int mxl_hostapd_ctrl_iface_receive_process(struct hostapd_data *hapd,
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 	} else if (os_strncmp(buf, "SEND_T2LM_TEARDOWN_FRAME ", sizeof("SEND_T2LM_TEARDOWN_FRAME ") - 1) == 0) {
 		*reply_len = mxl_hostapd_ctrl_iface_send_t2lm_teardown_frame(hapd, buf + sizeof("SEND_T2LM_TEARDOWN_FRAME ") - 1);
-	} else if (os_strncmp(buf, "DISASSOCIATE ", 13) == 0) {
-		mxl_mlo_hostapd_ctrl_disassociate(hapd, buf + 13, reply_len);
 	} else if (os_strncmp(buf, "ML_BSS_CRITICAL_UPDATE", 22) == 0) {
 		if (mxl_is_mlo_enabled(hapd->conf)) {
 			if (mxl_ml_handle_critical_update(hapd) < 0) {
diff --git a/src/ap/ctrl_iface_ap.c b/src/ap/ctrl_iface_ap.c
index df2e6f0d7..56b79cea3 100644
--- a/src/ap/ctrl_iface_ap.c
+++ b/src/ap/ctrl_iface_ap.c
@@ -674,6 +674,9 @@ int hostapd_ctrl_iface_disassociate(struct hostapd_data *hapd,
 	struct sta_info *sta;
 	const char *pos;
 	u16 reason = WLAN_REASON_PREV_AUTH_NOT_VALID;
+#ifdef CONFIG_IEEE80211BE_MXL_MLO
+	struct sta_info *sta1 = NULL;
+#endif
 
 	wpa_dbg(hapd->msg_ctx, MSG_DEBUG, "CTRL_IFACE DISASSOCIATE %s",
 		txtaddr);
@@ -724,10 +727,28 @@ int hostapd_ctrl_iface_disassociate(struct hostapd_data *hapd,
 
 	sta = ap_get_sta(hapd, addr);
 
+#ifdef CONFIG_IEEE80211BE_MXL_MLO
+	if (!sta && mxl_is_mlo_enabled(hapd->conf))
+		sta = mxl_ap_get_sta_by_non_ap_mld_mac(hapd, addr);
+#endif
+
 	if (os_strstr(txtaddr, " tx=0")) {
+#ifdef CONFIG_IEEE80211BE_MXL_MLO
+		if (sta) {
+			if (mxl_mlo_is_ml_sta(sta)) {
+				sta1 = sta->mxl_sta_info.linked_sta;
+				mxl_ml_stop_sta_traffic_remove_mld(hapd, sta);
+			}
+			ap_free_sta(hapd, sta);
+
+			if (sta1)
+				ap_free_sta(hapd->mxl_data.sibling_hapd, sta1);
+		}
+#else
 		hostapd_drv_sta_remove(hapd, addr);
 		if (sta)
 			ap_free_sta(hapd, sta);
+#endif
 #ifdef CONFIG_VENDOR_MXL
 	} else if (is_broadcast_ether_addr(addr) &&
 		(hapd->wpa_auth) && (hapd->conf->ieee80211w != NO_MGMT_FRAME_PROTECTION)) {
-- 
2.43.0

